DOCSP-47059: Work with BSON (#263)
mongodb/docs-php-library: source/data-formats/modeling-bson-data.txt

   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to create and interact with BSON documents
by using the {+driver-short+}.

**BSON**, or Binary JSON, is the data format that MongoDB uses to organize
and store data. This data format includes all JSON data structure types and

> supports additional types, including dates, different-sized integers, ``ObjectId``

Reviewer: S: to resolve vale error
```suggestion
also supports other types, such as dates, different-sized integers, ``ObjectId``
```

values, and binary data. The {+library-short+} provides the ``MongoDB\Model\BSONArray``
and ``MongoDB\Model\BSONDocument`` types to store BSON data.

.. tip::

>    For a complete list of supported BSON types, see :manual:`BSON Types
>    </reference/bson-types>` in the {+mdb-server+} manual.

Reviewer: ```suggestion
   To view a complete list of supported BSON types, see :manual:`BSON Types
   </reference/bson-types>` in the {+mdb-server+} manual.
```

.. _php-bson-sample:

Sample Data
~~~~~~~~~~~

The code examples in this guide reference the following sample BSON document:

.. code-block:: js

   {
       "address" : {
           "street" : "Pizza St", 
           "zipcode" : "10003" 
       },
       "coord" : [-73.982419, 41.579505]
       "cuisine" : "Pizza",

>        "name" : "Mongo's Pizza"

Reviewer: S: consider using a name that doesn't include a char that needs escaping, unless you specifically want to demonstrate this (which is good too!)

Author: Updated, I don't think it's necessary to demonstrate

   }

Create a BSON Document
----------------------

> You can create a BSON document by using the same notation you use to create an

Reviewer: S: [style guide](https://www.mongodb.com/docs/meta/style-guide/writing/use-that-which-correctly/)
```suggestion
You can create a BSON document by using the same notation that you use to create an
```

associative array in {+language+}. The {+library-short+} automatically converts
these values into BSON documents when inserting them into a collection.

The following example creates a BSON document that represents the preceding
:ref:`sample BSON document <php-bson-sample>`:

.. literalinclude:: /includes/bson.php
    :language: php
    :dedent:
    :start-after: start-create-doc
    :end-before: end-create-doc

Change a BSON Document
----------------------

> You can modify the contents of a BSON document by using the same notation you use to
> modify an associative array in {+language+}. This example makes the following changes
> to the :ref:`sample BSON document <php-bson-sample>`:

Reviewer: ```suggestion
You can modify the contents of a BSON document by using the same notation that you use to
modify an associative array in {+language+}. This example makes the following changes
to the :ref:`sample BSON document <php-bson-sample>`:
```

- Adds a new ``restaurant_id`` field that has a value of ``12345``

> - Sets the value of the ``name`` field to ``"Mongo's Pizza Place"``

Reviewer: S: I think the wording is confusing here. This code example only edits the sample BSON document in memory, not in MongoDB (as the updateOne method would do). Perhaps you can make this clearer as this isnt the same as a CRUD operation

Author: I wasn't sure how to make that more clear from the list, but I added a note after the code example

.. literalinclude:: /includes/bson.php
    :language: php
    :dedent:
    :start-after: start-modify-doc
    :end-before: end-modify-doc

Customize BSON Serialization
----------------------------

The following sections describe how to configure your application's 
serialization of BSON data:

- :ref:`php-type-map`: Use the ``typeMap`` option to specify the default conversion
  between {+language+} types and BSON types

- :ref:`php-persistable-classes`: Use the ``MongoDB\BSON\Persistable`` interface to
  enable serialization

- :ref:`php-enums`: Use the ``bsonSerialize()`` and ``bsonUnserialize()`` methods to

>   specify serialization between backed enums and BSON values

Reviewer: S: add periods - [Style guide](https://www.mongodb.com/docs/meta/style-guide/style/lists/#list-that-provides-terms-and-more-information)

> .. _php-type-map:

Reviewer: S: consider adding bson to this anchor unless its preserved from a prev version of this doc
```suggestion
.. _php-bson-type-map:
```

Type Maps
~~~~~~~~~

You can set the ``typeMap`` option, which configures how BSON values are converted to
{+language+} values, at the following levels:

- ``MongoDB\Client``, which sets the *default for all operations* unless overridden
- ``MongoDB\Database``
- ``MongoDB\Collection``

This list also indicates the increasing order of precedence of the option settings. For
example, if you set a ``typeMap`` for a collection, it will override the type map
set on the database.

The {+library-short+} uses the following type map by default:

.. code-block:: php

   [
       'array' => 'MongoDB\Model\BSONArray',
       'document' => 'MongoDB\Model\BSONDocument',
       'root' => 'MongoDB\Model\BSONDocument',
   ]

This type map performs the following conversions:

- Arrays to ``MongoDB\Model\BSONArray`` objects

> - Top-level and embedded BSON documents to ``MongoDB\Model\BSONDocument`` objects

Reviewer: S: clarify that these apply in both directions (serialization & deserialization)

Reviewer: This feedback applies to wording in other parts of this section as well

> A type map can specify any class that implements
> :php:`MongoDB\BSON\Unserializable <mongodb-bson-unserializable>`. It can

Reviewer: ```suggestion
A type map can specify any class that implements the
:php:`MongoDB\BSON\Unserializable <mongodb-bson-unserializable>` interface. It can
```

> also specify the ``array``, ``stdClass``, and ``object`` types.

Reviewer: ```suggestion
also specify conversions of the ``array``, ``stdClass``, and ``object`` types.
```

Custom Type Map Example
```````````````````````

The following example sets the ``typeMap`` option for the ``restaurants`` collection
that serializes arrays and BSON documents as ``MongoDB\Model\BSONDocument`` objects:

.. literalinclude:: /includes/bson.php
    :language: php
    :dedent:
    :start-after: start-type-map
    :end-before: end-type-map

.. _php-persistable-classes:

Persistable Classes
~~~~~~~~~~~~~~~~~~~

You can create classes that implement the ``MongoDB\BSON\Persistable`` interface,
which instructs the {+library-short+} to automatically perform serialization and
deserialization according to the {+extension-short+}'s :php:`persistence specification
<mongodb.persistence>`. The ``Persistable`` interface is analogous to {+language+}'s
:php:`Serializable interface <class.serializable>`.

When deserializing a PHP variable from BSON, the encoded class name of a
``Persistable`` object overrides any class specified in the ``typeMap`` option.
However, it does not override ``array``,  ``stdClass``, or ``object`` types.

.. tip::

   To learn more about the ``Persistable`` interface, see :php:`Persistable <mongodb-bson-persistable>`

>    in the {+extension-short+} documentation.

Reviewer: S: Instead of adding this admonition, inline the link when you first mention `Persistable`

Example
```````

> Consider the following ``Person`` class definition:

Reviewer: S: explain what is special/expected by defining the class this way

.. literalinclude:: /includes/bson.php
    :language: php
    :dedent:
    :start-after: start-person-class
    :end-before: end-person-class

The following example constructs a ``Person`` object, inserts it into the
database, and reads it back as an object of the same type:

.. io-code-block::
   :copyable:

   .. input:: /includes/bson.php
      :start-after: start-person-serialize
      :end-before: end-person-serialize
      :language: php
      :dedent:

   .. output::
      :visible: false

      object(Person)#18 (3) {
          ["id":"Person":private]=>
          object(MongoDB\BSON\ObjectId)#15 (1) {
          ["oid"]=>
          string(24) "56fad2c36118fd2e9820cfc1"
          }
          ["name":"Person":private]=>
          string(3) "Bob"
          ["createdAt":"Person":private]=>
          object(MongoDB\BSON\UTCDateTime)#17 (1) {
          ["milliseconds"]=>
          int(1459278531218)
          }
      }

The returned document is equivalent to the following BSON document:

> .. code-block:: js

Reviewer: Q: can you explain the significance of the `__pclass` field?

associative array in {+language+}. The {+library-short+} automatically converts
these values into BSON documents when inserting them into a collection.

The following example creates a BSON document that represents the preceding
:ref:`sample BSON document <php-bson-sample>`:

.. literalinclude:: /includes/bson.php
    :language: php
    :dedent:
    :start-after: start-create-doc
    :end-before: end-create-doc

Change a BSON Document
----------------------

You can modify the contents of a BSON document by using the same notation that you use to
modify an associative array in {+language+}. This example makes the following changes
to the :ref:`sample BSON document <php-bson-sample>`:

- Adds a new ``restaurant_id`` field that has a value of ``12345``
- Changes the ``name`` field value to ``"Galaxy Pizza"``

.. literalinclude:: /includes/bson.php
    :language: php
    :dedent:
    :start-after: start-modify-doc
    :end-before: end-modify-doc

.. note::

   The preceding code changes only the in-memory values of the sample BSON
   document. It does not run any database operations that change values stored
   in MongoDB. To learn how to modify documents stored in MongoDB, see the

>    :ref:`php-write-update` guide.

Reviewer: This looks good and addresses my concern!

You can serialize and deserialize backed enums into BSON data. Backed 
enum values serialize as their case value, while pure enums without
case values cannot be directly serialized. To perform these conversions,
you must specify serialization logic by defining the ``bsonSerialize()``
and ``bsonUnserialize()`` methods in your class definition.

.. tip::

   To learn more about backed enums, see :php:`Backed enums <enumerations.backed>`
   in the {+extension-short+} documentation.

Example
```````

Consider the following ``User`` class definition, which specifies
logic for serializing and deserializing its fields into BSON values.
The class includes a ``role`` field, which has a backed enum value:

.. literalinclude:: /includes/bson.php
    :language: php
    :emphasize-lines: 15, 24
    :dedent:
    :start-after: start-backed-enum
    :end-before: end-backed-enum

The following example constructs a ``User`` object with a ``role`` field,
inserts it into the database, and reads it back as an object of the same type:

.. io-code-block::
   :copyable:

   .. input:: /includes/bson.php
      :start-after: start-enum-serialize
      :end-before: end-enum-serialize
      :language: php

>       :dedent:

Reviewer: S: maybe its worth adding a sentence below here that says that the USER enum value was stored as a 1 in MongoDB, but deserialized correctly into the enum value. But can also leave out if its obvious/redundant.

.. note::

   You can use the ``Persistable`` interface for root and embedded BSON documents
   only, not BSON arrays.

.. _php-enums:

Enum Values
~~~~~~~~~~~

You can serialize and deserialize backed enums into BSON data. Backed 
enum values serialize as their case value, while pure enums without
case values cannot be directly serialized. To perform these conversions,
you must specify serialization logic by defining the ``bsonSerialize()``
and ``bsonUnserialize()`` methods.

.. tip::

   To learn more about backed enums, see :php:`Backed enums <enumerations.backed>`
   in the {+extension-short+} documentation.

Example
```````

The following example specifies logic for serializing and deserializing the
``Role`` enum by defining the ``bsonSerialize()`` and ``bsonUnserialize()``
methods:

.. literalinclude:: /includes/bson.php
    :language: php
    :dedent:
    :start-after: start-backed-enum

>     :end-before: end-backed-enum

Reviewer: S: consider adding highlighting for the methods of interest

> 

Reviewer: S: consider demonstrating an example as you did in the previous section of the serialization roundtrip

.. note::

>    Enums cannot implement the ``MongoDB\BSON\Unserializable`` and
>    ``MongoDB\BSON\Persistable`` interfaces, since enum cases have no
>    state and cannot be instantiated like class objects. However, pure and backed
>    enums can implement ``MongoDB\BSON\Serializable``, which you can use to
>    override the default enum serialization behavior.

Reviewer: ```suggestion
   Enums cannot implement the ``MongoDB\BSON\Unserializable`` and
   ``MongoDB\BSON\Persistable`` interfaces because enum cases have no
   state and cannot be instantiated like class objects. However, pure and backed
   enums can implement ``MongoDB\BSON\Serializable``, which you can use to
   override the default enum serialization behavior.
```
